### PR TITLE
Add TLS to gRPC server

### PR DESCRIPTION
**Pull request by aloksharma20**
- Added new configuration fields to enable TLS for gRPC server.

### CODE REVIEW

File: src/conf/local-file.yaml [hunk 1]
```diff
@@ -131,3 +131,12 @@ pprof: false
...
   auto-deploy-dsp: false
+
+# Feeder Service 
+feederservice:
+  enable: true
+  tls:
+    cert:
+      file: /home/alok-pc/go/src/github.com/aloksharma20/discovery-engine/src/cert/ca-cert.pem
+    key:
+      file: /home/alok-pc/go/src/github.com/aloksharma20/discovery-engine/src/cert/ca-key.pem
```
Comment (wazir-ahmed) [inline src/conf/local-file.yaml]:
```suggestion
server:
  port: 9089
  tls:
    enable: false
    cert: /tls/cert.pem
    key: /tls/key.pem
```
- Let's use this structure for the configuration.
- Currently we have hardcoded the server port in `grpcServer.go`. Let's make it configurable.

File: src/server/grpcServer.go [hunk 2]
```diff
@@ -362,3 +370,13 @@ func AddServers(s *grpc.Server) *grpc.Server {
 
 	return s
 }
+
+func GetTLSCredentails() credentials.TransportCredentials {
+	certFile := viper.GetString("feederservice.tls.cert.file")
+	keyFile := viper.GetString("feederservice.tls.key.file")
+	creds, err := credentials.NewServerTLSFromFile(certFile, keyFile)
+	if err != nil {
+		log.Error().Msgf("Unable to read tls certificate credentails ::  %s", err)
```
Comment (wazir-ahmed) [inline src/server/grpcServer.go:382]:
```suggestion
		log.Error().Msgf("Unable to read tls certificate credentails ::  %s", err)
		return nil
```

**Review by wazir-ahmed — CHANGES_REQUESTED** [review]
Things that has to added.
1. The configuration changes has to added to deployment manifest and helm configuration.
2. Needs to handle certificate auto-generation.

File: src/server/grpcServer.go (hunk 2 above)
Comment (aloksharma20) [reply src/server/grpcServer.go:382]: addressed

File: src/server/grpcServer.go [hunk 3]
```diff
@@ -303,7 +303,18 @@ func (ps *publisherServer) GetSummary(req *ppb.SummaryRequest, srv ppb.Publisher
...
 
 func StartGrpcServer() *grpc.Server {
-	s := grpc.NewServer()
+	var s *grpc.Server
+	if viper.GetBool("server.tls.enable") {
+		creds := GetTLSCredentails()
+		if creds != nil {
+			s = grpc.NewServer(grpc.ServerOption(grpc.Creds(creds)))
+		} else {
+			log.Error().Msgf("Unable to read credentails ::  %s", creds)
```
Comment (wazir-ahmed) [inline src/server/grpcServer.go]:
```suggestion
			log.Fatal().Msgf("Unable to read credentails ::  %s", creds)
```
It should be `Fatal`. In the program reaches this error state, it cannot proceed after that.

File: src/main.go [hunk 4]
```diff
@@ -63,6 +63,7 @@ func init() {
 	log.Info().Msgf("SYSTEM-POLICY: %+v", config.GetCfgSys())
 	log.Info().Msgf("KUBEARMOR: %+v", config.GetCfgKubeArmor())
 	log.Info().Msgf("AUTO-DEPLOY-DSP: %+v", config.GetCfgDsp())
+	log.Info().Msgf("is TLS enable for feeder service :: %t", viper.GetBool("server.tls.enable"))
```
Comment (wazir-ahmed) [inline src/main.go]:
```suggestion
	log.Info().Msgf("TLS enabled: %t", viper.GetBool("server.tls.enable"))
```

**Review by wazir-ahmed — CHANGES_REQUESTED** [review]

File: src/server/grpcServer.go (hunk 3 above)
Comment (aloksharma20) [reply src/server/grpcServer.go]: addressed

File: src/main.go (hunk 4 above)
Comment (aloksharma20) [reply src/main.go]: addressed

**Review by wazir-ahmed — APPROVED** [review]

File: src/server/grpcServer.go [hunk 5]
```diff
@@ -303,7 +303,18 @@ func (ps *publisherServer) GetSummary(req *ppb.SummaryRequest, srv ppb.Publisher
...
+	if viper.GetBool("server.tls.enable") {
+		creds := GetTLSCredentails()
+		if creds != nil {
+			s = grpc.NewServer(grpc.ServerOption(grpc.Creds(creds)))
+		} else {
+			log.Fatal().Msgf("Unable to read credentails ::  %s", creds)
+		}
+	} else {
+		s = grpc.NewServer()
+	}
```
Comment (avithe-great) [inline src/server/grpcServer.go:316]:
```
var opts []grpc.ServerOption

	if viper.GetBool("server.tls.enable") {
		creds := GetTLSCredentials()
		if creds != nil {
			opts = append(opts, grpc.Creds(creds))
		} else {
			log.Fatal().Msg("Unable to read TLS credentials")
		}
	}

	s := grpc.NewServer(opts...)
```
Looks good to me. but shall we handle the options this way. @aloksharma20 ?

**Review by avithe-great — APPROVED** [review]

**Comment by wazir-ahmed** [comment]
Merging this since this is been pending for a long time.

**Pull request merged**